README.md
nodeserver
==========

Production webserver for Node.js applications, php websites, statics websites, etc.

![nodeserver logo](https://raw.githubusercontent.com/altairstudios/nodeserver/master/nodeserver-logo.png)

Nodeserver is a production webserver Node.js applications. It allows you to keep applications alive forever, to reload them, start webistes on php or html, reverse proxy.

If you have ideas or questions, open a new issue with your ideas.

[![Version npm](https://img.shields.io/npm/v/nodeserver.svg?style=flat-square)](https://www.npmjs.com/package/nodeserver)[![NPM Downloads](https://img.shields.io/npm/dm/nodeserver.svg?style=flat-square)](https://www.npmjs.com/package/nodeserver)[![Travis Build Status](https://api.travis-ci.org/altairstudios/nodeserver.svg)](https://travis-ci.org/altairstudios/nodeserver)



## How install

Install nodeserver script via npm:

```bash
$ sudo npm install nodeserver -g
```

Install daemon on the system. Currently only support CentOS.

```bash
$ sudo nodeserver install centos
```


## Configuration

By default nodeserver read configuration in /etc/nodeserver/nodeserver.config

You can use other path using the command:

```bash
$ nodeserver start /path/to/config
```

or use loop nodeserver with

```bash
$ nodeserver start-loop /path/to/config
```


### Basic configuration

The basic configuration file are json with format:

```json
{
  "nodeserver": {
    "admin": {
      "active": "true|false - active the admin web interface",
      "port": "admin listen port. We recommend use port 10000",
      "user": "user for admin access, we recommend not use admin, root or similars",
      "password": "hash password. See generate password section"
    }
  },
  "sites": []
}
```

The admin interface is a web page where administrator can start, stop process, view logs, etc. If active is false, the other options not mandatory.


### Site configuration

Sites is a array of sites. The site has the next format:

```json
{
  "name": "Site name, identify in administrator cli or web",
  "type": "Site type use supported workers: cdn, cgi, php, node, python",
  "bindings": ["array of string with domain (or ip) and port. By default use domain and port 80: example.com:80"],
  "port": "use only in node site. Start the node site in theese port",
  "script": "absolute path of script server in node sites or document root in php, cgi o cdn sites",
  "security": {
    "certs": {
      "key": "absolute path for certificate key",
      "cert": "absolute path for certificate",
      "ca": ["array of certificate ca"]
    },
    "bindings": [
      "similar to bindings option but uses only for SSL connection. By default use domain with port 443, example.com:443"
    ]
  }
}
```

The port option is only mandatory for nodejs sites.

Use the security option only if you use a SSL in sites.


### Reference site configuration

In config file you can use a string for any site where determine the relative path for configuration site. If use /etc/nodeserver for configuration you can create a "sites" folder inside and save a example.config for save json configuration for this site.

A simple configuration nodeserver.config is:

```json
{
  "nodeserver": {
    "admin": {
      "active": false
    }
  },
  "sites": [
    "sites/example.com"
  ]
}
```

The site / example.com.config file are:

```json
{
  "name": "Example site",
  "type": "cdn",
  "bindings": [
    "example.com:80"
  ],
  "script": "/var/web/example.com"
}
```


### Servers operations

Start server

```bash
  $ sudo nodeserver start
```

Check status and show a list of active webites

```bash
  $ nodeserver status
```

![nodeserver status](https://raw.githubusercontent.com/altairstudios/nodeserver/master/images/status.png)

Stop the server

```bash
  $ nodeserver stop
```

Reload server configuration if nodeserver.config change without stop the server

```bash
  $ nodeserver reload
```

Generate a hash password for use in config file

```bash
  $ nodeserver password mypasswordtohash
```

## Thanks

NodeServer is a free and open source community-driven project.

Thanks to the following companies and projects whose work we have used or taken inspiration from in the making of NodeServer:

* [AltairStudios](http://www.altairstudios.es)
* [Node.js](http://www.nodejs.org)
* [KeystoneJS](http://www.keystonejs.com)
* [ExpressJS](http://www.expressjs.com)
* [Digital Ocean](http://www.digitalocean.com/)



## License

(The MIT License)

Copyright (c) 2014-2017 Juan Benavides

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software'), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


HISTORY.md
nodeserver History Version
==========================


## Lasts versions

### v0.9

#### v0.9.1

* Update Travis with last nodejs versions
* New CDN site with plain html format

#### v0.9.0

* Update Travis with last nodejs versions
* Add multiple ssl for a single process using security.custom[].bingings

### v0.8

#### v0.8.3

* Fixed tests warnings

#### v0.8.2

* Fixed ".+:80" domain pattern

#### v0.8.1

* Fixed SSL problem for differents ports as 80 and 443.
* Stop funtionality add stop website child proccess before exit main proccess

#### v0.8.0

* Fixed error when refresh a webserver with no nodejs websites (php, cgi)
* Split separated workers type in files for node, cgi, php, python (not implemented yet)
* Updated travis test to nodejs 5.5
* #18 Show in the admin interface a log with attempt login on admin 
* Add support to separate configuration in files using string object against json object. The file is separated in folder "/sites/name.config" 

### v0.7

#### v0.7.3

* Fixed error when stop server or refresh server config not stop the childs nodejs processes and duplicate them.

#### v0.7.2

* Fixed error when try exec "nodeserver status" and server down by the error

#### v0.7.1

* Fixed problem when close the server process
* Fixed admin close interface
* Fixed an error when admin port is used
* Added more admin tests
* Updated admin to express4
* Set force detached false for childproccess of nodejs
* Added travis tests for nodejs 5.1.0

#### v0.7.0

* Clean nodeserver bash script
* Change nodeserver script init from instance class to a class for initialize with a param 'inTerminal'. Example new require('nodeserver')(false)
* Check in terminal if socket exits but not a server listen, kill this socket and show a message 'server not running'
* Updated the mocha tests at last version
* Update package version to new version v0.7.0
* Add 6 mocha configuration tests
* Add stats from NPM on README.md file
* Added mocha to script package and added travis configuration
* Some fixes over nodejs 0.10.x with SSL
* Fixed socket close error when process exit on nodejs 0.10.40 - throw new Error('Not running');
* Added gulp for future build and lint to test
* Fixed lint errors

### v0.6

#### v0.6.3

* Fixed an error with regex domains bindings

#### v0.6.2

* Fixed an error with regex domains bindings

#### v0.6.1

* Add status information on shell $ nodeserver status
* Add example on readme of status output

#### v0.6.0

* Change not respond websites from 400 to 504 HTTP code
* Deleted reboot button on admin. Now use stop and start buttons
* Refactored website model
* Refactored worker for nodejs to external workers file with cgi worker
* Added log in admin with the last 100 log lines

### v0.5

#### v0.5.4

* New worker class for manage all request for nodejs and cgi.
* Add a new method getWebsiteFromBinding instead of getWebsiteFromUrl and getWebsiteFromSecureUrl

#### v0.5.3

* Added a 400 error when a site is down
* Fixed a error that shutdown server when a cgi site port is over 65536
* Deleted old xml mimes and added all to mimes.json

#### v0.5.2

* Updated documentation of install and use

#### v0.5.1

* Fixed some errors on install system

#### v0.5.0

* Added terminal daemon for start
* Added terminal daemon for status
* Added terminal daemon for stop
* Change website id format
* Encrypted password for access to the admin
* Refresh without stop the server all websites
* Change absoluteScript configuration to optionaly and mark false by default
* Add installation system for CentOS

### v0.4

#### v0.4.0

* Added support for ssl. New configuration security json for specify certificates and ssl binding (443 for default).

### v0.3

#### v0.3.2

* fixed a error when uses a root domain and subdomians. When use domain like demo.com with a home website and services.demo.com the regex pattern resolve the root domain demo.com when uses http://services.demo.com.

#### v0.3.1

* Added server logo to admin website
* Added support to wildcard with regex to bindings domains

#### v0.3.0

* New admin interface
* Fixed many bugs

### v0.2

#### v0.2.0

* New config with sites and nodeserver config
* Admin interface configuration

### v0.1

#### v0.1.2

* Server file .js and package.json are watched when change restar this node server

#### v0.1.1

* Fixed a error in CentOS node install

#### v0.1.0

* Daemon funtionality with config file and forever monitor

nodeserver.js
var httpProxy = require('http-proxy');
var http = require('http');
var https = require('https');
var agent = new http.Agent({ maxSockets: Number.MAX_VALUE });
var proxy = httpProxy.createProxy({ xfwd: true, agent: agent });
var fs = require('fs');
var nodeserverAdmin = require('./admin');
var os = require('os');
require('colors');
var core = require('./core');
var crypto = require('crypto');
var tls = require('tls');
var path = require('path');



module.exports = exports = function(inTerminal) {
	var self = this;
	this.websites = [];
	this.ports = [];
	this.securePorts = [];
	this.baseCerts = {};
	this.servers = [];
	this.config = null;
	this.unix = (os.platform() == 'darwin' || os.platform() == 'linux');
	this.running = false;
	this.socket = null;
	this.configFile = '/etc/nodeserver/nodeserver.config';

	core.terminal.nodeserver = this;
	core.sockets.nodeserver = this;
	core.workers.nodeserver = this;
	core.workers.proxy = proxy;



	/**
	 * Stop all process of server, read again the configuration file and restart the server
	 * @return {null}
	 */
	this.hotConfig = function() {
		this.stop();
		this.readConfigFile(this.configFile);
		this.start(false);
	};



	/**
	 * Get version number of nodeserver
	 * @return {string}
	 */
	this.getVersion = function() {
		return require('./package.json').version;
	};



	this.getWebsiteFromBinding = function(url, onlySecure, ignorePort) {
		onlySecure = onlySecure || false;
		ignorePort = ignorePort || false;



		var containBinding = function(url, bindings, hasRegex) {
			var bindingsCount = bindings.length;

			for(var i = 0; i < bindingsCount; i++) {
				var binding = bindings[i];

				if(hasRegex) {
					var regex = new RegExp('^' + binding + '$', 'gi');
					
					if(regex.test(url)) {
						return true;
					}
				} else {
					if(binding == url) {
						return true;
					}
				}
			}

			return false;
		};


		var getWebsite = function(url, websites) {
			var websitesCount = websites.length;

			for(var i = 0; i < websitesCount; i++) {
				var website = websites[i];

				if(website.security) {
					var result = containBinding(url, website.security.bindings, false);
					if(result) return website;
				}

				if(!onlySecure) {
					var result = containBinding(url, website.bindings, false);
					if(result) return website;
				}
			}

			for(var i = 0; i < websitesCount; i++) {
				var website = websites[i];

				if(website.security) {
					var result = containBinding(url, website.security.bindings, true);
					if(result) return website;
				}

				if(!onlySecure) {
					var result = containBinding(url, website.bindings, true);
					if(result) return website;
				}
			}
		};


		var thewebsite = getWebsite(url, self.websites);

		if(thewebsite == null) thewebsite = getWebsite(url + ':80', self.websites);
		if(thewebsite == null) thewebsite = getWebsite(url + ':443', self.websites);
		
		if(thewebsite) {
			return thewebsite;
		}

		for (var i = self.securePorts.length - 1; i >= 0; i--) {
			var secureWebsite = getWebsite(url + ':' + self.securePorts[i], self.websites);
			if(secureWebsite != null) {
				return secureWebsite;
			}
		}

		return thewebsite;
	};



	this.readConfigFile = function(configFile) {
		configFile = configFile || 'nodeserver.config';
		var config = null;

		if(!path.isAbsolute(configFile)) {
			configFile = path.resolve(process.cwd(), configFile);
		}

		if(fs.existsSync(configFile)) {
			this.configFile = configFile;
			config = fs.readFileSync(configFile);
		} else if(fs.existsSync(__dirname + configFile)) {
			this.configFile = __dirname + configFile;
			config = fs.readFileSync(__dirname + configFile);
		}  else if(fs.existsSync(__dirname + '/' + configFile)) {
			this.configFile = __dirname + '/' + configFile;
			config = fs.readFileSync(__dirname + '/' + configFile);
		} else if(fs.existsSync('/etc/nodeserver/nodeserver.config')) {
			this.configFile = '/etc/nodeserver/nodeserver.config';
			config = fs.readFileSync('/etc/nodeserver/nodeserver.config');
		}

		var json = JSON.parse(config);

		this.readConfig(json);
	};



	this.readConfig = function(config) {
		this.config = config;

		if(this.config == null) {
			return;
		}

		this.websites = [];

		if(this.config.sites && Array.isArray(this.config.sites)) {
			for(var i = 0; i < this.config.sites.length; i++) {
				this.addWebsite(this.config.sites[i]);
			}	
		}
		

		if(this.config.nodeserver && this.config.nodeserver.admin) {
			if(this.config.nodeserver.admin.active) {
				if(!this.admin) {
					this.admin = new nodeserverAdmin(self);
					this.admin.adminInterface();
				}
			} else if(this.admin) {
				this.admin.stopAdminInterface();
			}	
		}
	};



	this.addWebsite = function(site) {
		var website = new core.models.website(site, this);

		for(var i = 0; i < website.ports.http.length; i++) {
			this.addPort(website.ports.http[i]);
		}

		for(var i = 0; i < website.ports.https.length; i++) {
			this.addSecurePort(website.ports.https[i]);
		}

		if(website.security.certs != null && website.security.certs != undefined && website.security.certs.key != null) {
			this.baseCerts = website.security.certs;
		}

		this.websites.push(website);
		core.workers.start(website);
	};



	this.addPort = function(port) {
		for(var i = 0; i < this.ports.length; i++) {
			if(this.ports[i] == port) {
				return;
			}
		}

		this.ports.push(port);
	};


	this.addSecurePort = function(port) {
		for(var i = 0; i < this.securePorts.length; i++) {
			if(this.securePorts[i] == port) {
				return;
			}
		}

		this.securePorts.push(port);
	};



	this.start = function(started) {
		if(started === undefined) {
			if(this.unix) {
				core.sockets.start();

			}
		}

		var ports = this.ports.length;
		var securePorts = this.securePorts.length;

		for(var i = 0; i < ports; i++) {
			var server = http.createServer(core.workers.process);
			server.listen(this.ports[i]);

			this.servers.push(server);
		}


		var secureOptions = {
			SNICallback: function(domain, callback) {
				var website = self.getWebsiteFromBinding(domain, true, true);

				if(website) {
					var security = website.getSecurity(domain);
					/*{
						key: fs.readFileSync(website.security.certs.key),
						cert: fs.readFileSync(website.security.certs.cert)
					};*/

					if(website.security.certs.ca) {
						security.ca = [];

						for (var i = website.security.certs.ca.length - 1; i >= 0; i--) {
							security.ca.push(fs.readFileSync(website.security.certs.ca[i]));
						}
					}

					if(callback) {
						callback(null, tls.createSecureContext(security));
					} else {
						try {
							return tls.createSecureContext(security);
						} catch(e) {
							return crypto.createCredentials(security).context;
						}
					}
				} else {
					if(callback) {
						callback(true);
					} else {
						return true;
					}
				}
			},
			key: (self.baseCerts.key && fs.existsSync(self.baseCerts.key)) ? fs.readFileSync(self.baseCerts.key) : '',
			cert: (self.baseCerts.cert && fs.existsSync(self.baseCerts.cert)) ? fs.readFileSync(self.baseCerts.cert) : ''
		};

		for(var i = 0; i < securePorts; i++) {
			var server = https.createServer(secureOptions, core.workers.process);
			server.listen(this.securePorts[i]);
			this.servers.push(server);
		}

		this.running = true;
	};



	this.stop = function() {
		for(var i = 0; i < this.websites.length; i++) {
			if(this.websites[i].operations) {
				this.websites[i].operations.stop();
			}
		}

		for(var i = 0; i < this.servers.length; i++) {
			this.servers[i].close();
		}
	};


	this.restart = function() {
		this.stop();
		this.start();
	};



	this.exit = function() {
		this.stop();

		if(this.admin) {
			this.admin.stopAdminInterface();
		}

		if(self.unix && self.socket) {
			try {
				self.socket.close();
			} catch(ex) {}
		}
	};



	//process.setMaxListeners(100);



	process.on('exit', function() {
		self.exit();
	});




	/*process.on('uncaughtException', function(err) {
		/*eslint-disable*-/
		console.error('Nodeserver Error ' + err);
		console.error(arguments);
		/*eslint-enable*-/
	});*/


	process.on('SIGINT', function() {
		/*eslint-disable*/
		console.log('\nSayonara baby!!');
		/*eslint-enable*/
		process.exit(0);
	});


	this.terminal = core.terminal.process;

	if(inTerminal) {
		this.terminal(process.argv);
	}

	return self;
};